pravega/pravega
# Issue 4231: Use system truststore for TLS when client truststore is not specified

**Change log description**  
Enable use of system truststore for client-server TLS communications when client truststore is unspecified.

**Purpose of the change**  
Fixes #4231. 

**What the code does**  
Here's what the code does: 
* Removes the use of `FingerprintTrustManagerFactory` when client truststore is null or empty. 
* Instead, sets Netty `SslContextBuilder trustManager` as null, so that system truststore is used 

**How to verify it**  
All unit and integration tests should pass. A couple of focused tests used for testing the PR are as follows: 
1. Verify that clients are able to read/write events over TLS-enabled channels when client truststore is specified. 
   Run tests in `TlsEnabledInProcPravegaClusterTest` for the purpose. 
2. Verify that client are able to read/write events when the CA certificate is not specified via `ClientConfig` but is present in the system truststore. 
   1. Create new TLS material for Pravega standalone using steps described [here](https://github.com/pravega/pravega/wiki/Creating-TLS-material-for-Pravega-security). 
   2. Modify Pravega standalone config to enable TLS and Auth. Point TLS params to newly created material. Here's how my configuration looked like: 
      ```
      singlenode.enableAuth=true
      singlenode.userName=admin
      singlenode.passwd=1111_aaaa
      singlenode.passwdFile=../config/passwd

      singlenode.enableTls=true
      singlenode.keyFile=/home/xyz/temp/pki-client-truststore/server-key.key
      singlenode.certFile=/home/xyz/temp/pki-client-truststore/server-cert.crt
      singlenode.keyStoreJKS=/home/xyz/temp/pki-client-truststore/server.keystore.jks
      singlenode.keyStoreJKSPasswordFile=/home/xyz/temp/pki-client-truststore/server.keystore.jks.passwd
      ```
   3.  Run a reader/writer client against the standalone server (like the one shown [here](https://github.com/pravega/pravega-samples/tree/master/pravega-client-examples/src/main/java/io/pravega/example/secure), without specifying the client truststore. You should see an error like the following on the client-side: 
 ```
       2019-09-30 23:13:19.467 [grpc-default-executor-1] WARN  i.p.c.stream.impl.ControllerImpl.warn(285) - gRPC call for createScope with trace id 3903548272740574148 failed with server error.
io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
	at io.grpc.Status.asRuntimeException(Status.java:530)
at io.grpc.stub.ClientCalls$StreamObs
      ....
     Caused by: javax.net.ssl.SSLHandshakeException: General OpenSslEngine problem
	at io.netty.handler.ssl.ReferenceCountedOpenSslContext$AbstractCertificateVerifier.verify(ReferenceCountedOpenSslContext.java:629)
	at io.netty.internal.tcnative.SSL.readFromSSL(Native Method)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.readPlaintextData(ReferenceCountedOpenSslEngine.java:511)
...
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:397)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:302)
	at sun.security.validator.Validator.validate(Validator.java:262)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:330)
at sun.sec       

```
   4. Now, add the CA certificate to the system truststore: 
       ```
        $ sudo keytool -import -trustcacerts -keystore /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/security/cacerts -storepass changeit -alias newlocalhost -import -file ca-cert.crt
       ```
  5. Now, run the reader/writer test again. This time the tests should succeed.
